### PR TITLE
[SPARK-47522][SQL] Read MySQL FLOAT as FloatType to keep consistent with the write side

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Read MySQL FLOAT as `FloatType` to keep consistent with the write side. In `getJDBCType`, we map `FloatType` to `JdbcType("FLOAT", java.sql.Types.FLOAT)`. So, in `getCatalystType`, we shall keep it consistent.

### Why are the changes needed?

read/write roundtrip for MySQL float

### Does this PR introduce _any_ user-facing change?

yes, please refer to the migration doc

### How was this patch tested?
new tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 1]
```diff
@@ -123,6 +123,8 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper {
         // Signed values in [-8388608, 8388607] and unsigned values in [0, 16777215],
         // both of them fit IntegerType
         Some(IntegerType)
+      case Types.REAL | Types.FLOAT =>
+        if (md.build().getBoolean("isSigned")) Some(FloatType) else Some(DoubleType)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:127]: Just a question. We are safe even if MySQL community remove UNSIGNED FLOAT , right?
- https://dev.mysql.com/doc/refman/8.3/en/numeric-type-syntax.html
> The UNSIGNED attribute is deprecated for columns of type [FLOAT](https://dev.mysql.com/doc/refman/8.3/en/floating-point-types.html) (and any synonyms) and you should expect support for it to be removed in a future version of MySQL.

Although it's a separate question, do you think we need to warn from our side, @yaooqinn ?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -938,9 +938,17 @@ class JDBCSuite extends QueryTest with SharedSparkSession {
...
+      Some(FloatType))
+    assert(mySqlDialect.getCatalystType(java.sql.Types.FLOAT, "FLOAT", 1, metadata) ===
+      Some(FloatType))
     metadata.putBoolean("isSigned", value = false)
     assert(mySqlDialect.getCatalystType(java.sql.Types.TINYINT, "TINYINT", 1, metadata) ===
       Some(ShortType))
+    assert(mySqlDialect.getCatalystType(java.sql.Types.REAL, "FLOAT", 1, metadata) ===
+      Some(DoubleType))
+    assert(mySqlDialect.getCatalystType(java.sql.Types.FLOAT, "FLOAT", 1, metadata) ===
+      Some(DoubleType))
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:951]: If we use the following, do you know when `DOUBLE` is required?

https://dev.mysql.com/doc/refman/8.3/en/floating-point-types.html
> FLOAT[(M,D)] [UNSIGNED] [ZEROFILL]

- A small (single-precision) floating-point number.

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:951]: Hi @dongjoon-hyun, this is a good point.

A `float(p)` can represent a single-precision float when p <= 23 and double-precision when p > 23 in MySQL. This is already handled by mysql-connect/j, so we get java.sql.Types.FLOAT/REAL for the former and java.sql.Types.DOUBLE for the latter. We do not need to worry about the case that we use our `FloatType` to get a MySQL Float(p>23).

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:951]: Thank you for the clarification.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:127]: Hi @dongjoon-hyun. Yes, it's safe, we get `isSigned` from the JDBC standard API. It's OK for the DB side to remove this.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0~

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:127]: > do you think we need to warn from our side

Oh, I missed this. This won't be necessary, it looks better to be warned when users use DDLs against MySQL DB. In this part, we actually act as a reader

**Comment by yaooqinn** [comment]
Opps, I added more tests for float(p) but didn't get a chance to upload it.:)

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:951]: You are welcome @dongjoon-hyun

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun. I sent #45672 a follow-up to improve the test coverage.

**Pull request closed without merging**